nixos/github-runners: remove single runner module, rework default name, add `noDefaultLabels`

veehaitch opened the pull request:
## Description of changes

Reworks the NixOS module in a few ways:

### Remove `services.github-runner`

Before #176691, the only way to configure a single GitHub Actions Runner was through the module `services.github-runner`.
Now that we have `services.github-runners.<name>` which allows configuration of multiple runners, there's no need to keep the old module around.                                                                                                               The PR removes `services.github-runner` through `mkRemovedOptionModule` which points the user to `services.github-runners.<name>`.
I also added a release notes entry.

### Default runner name

So far, the registration used `networking.hostName` if the configuration does not set `name`.
However, not setting `name` was only possible through `services.github-runner` which this PR removes; when using `services.github-runners.*` the `name` defaults to the attribute name.
Since the runner configuration script defaults to the hostname if the `--name` option is not used, the PR now allows setting `name = null` but still defaults to the attribute name.

### Add module option `noDefaultLabels`

[v2.305.0](https://github.com/actions/runner/releases/tag/v2.305.0) added the configuration option `--no-default-labels` which our module didn't cover so far.
The PR adds the module option `noDefaultLabels` and asserts that the configuration uses `extraLabels` if the option is set to `true`.
### Documentation / Node runtime / Cosmetics

Make the option documentation reflect that changing the `workDir` and the `ephemeral` triggers a new runner registration.
The module option `nodeRuntimes` still accepted `node16` as a value which the package does not accept anymore (since we don't have Node.js 16 in nixpkgs anymore).                                                                                             Also removed some superfluous usages of `lib`.

### Adapt `user` and add `group`

While the documentation tells us that the service would only run as a systemd dynamic user if `user == null`, this was never the case. Even if someone set the `user` option, this would only influence the username but would still use a dynamically allocated user. This PR makes the service reflect the documentation.

Also adds a new `group` option which complements the already existing `user` option. The effect of the `group` option also depends on the value `user` is set to. The option's documentation describes the behavior in detail.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

veehaitch committed:
nixos/github-runners: remove superfluous usages of `lib`

veehaitch committed:
nixos/github-runners: add `noDefaultLabels` option

Add option `noDefaultLabels` which controls the `--no-default-labels`
switch passed to the configure script.

veehaitch committed:
nixos/github-runners: document new runner registrations

Make the documentation of the options `workDir` and `ephemeral` reflect
that changing these options will trigger a new runner registration.

veehaitch committed:
nixos/github-runner: deprecate module in favor of `github-runners.*`

veehaitch committed:
nixos/github-runners: rework `name` default

veehaitch committed:
nixos/github-runners: align `nodeRuntimes` option with package

The `github-runner` package only supports `nodejs_20` since `nodejs_16`
was removed in a2976db919afc97b69ce3b7699a1a44ce61bbe5f.

It still makes sense to keep the `nodeRuntimes` option as this is
probably not the last Node.js we'll deprecate with at least some grace
period.

veehaitch commented:
@ofborg test github-runner

Trundle reviewed (approved):
The changes look good to me. I also checked
* that using the removed option points to using `services.github-runners.*`
* that the default `name` and `name = null` work
* setting `noDefaultLabels = true` requires `extraLabels` to be set

newAM reviewed (approved):
Have not tested it, but changes look good, a much needed cleanup, thanks!

aanderse commented on nixos/modules/services/continuous-integration/github-runner/options.nix:
```diff
@@ -1,213 +1,250 @@
-    description = lib.mdDoc ''
-      Modify the systemd service. Can be used to, e.g., adjust the sandboxing options.
-      See {manpage}`systemd.exec(5)` for more options.
+  config.assertions = flatten (
```
why is this in the `options.nix` file, which operates on `options`, instead of the `service.nix` file, which operates on `config`?

veehaitch replied:
Moved.

aanderse commented on nixos/modules/services/continuous-integration/github-runner/options.nix:
```diff
@@ -1,213 +1,250 @@
+          default = false;
+        };
+
+        user = mkOption {
```
not related to this PR, but it jumped out at me again while looking at this...

the documentation is incorrect here - this service will *always* use a `DynamicUser`, even if you set a value here

can we possibly update the code to reflect the documentation, or the documentation to reflect the code? specifically if the user sets a value here then `DynamicUser` should be `false` **or** mention that if the user sets a value here there will still be a `DynamicUser`, but the dynamic user will have a name

i would very much appreciate if i could set a value in this option and actually have a real user instead of dynamic

veehaitch replied:
Yes, you're right and I changed the systemd service configuration to not use a dynamic user if `user != null`. I also added a `group` option to complement the already existing `user` option.

aanderse commented:
unfortunate for me... but this PR breaks my workflow :disappointed:

the `systemd` hardening in this module just doesn't work for my usage and there isn't an easy way to undo it with this PR
currently [this](https://github.com/aanderse/nix/blob/main/nixos/modules/github-runner/default.nix) is what i'm doing... though what i would really love is a way to simply strip out the `systemd` hardening in this module and then i wouldn't need to resort to the hacks i use in the link provided

@veehaitch given the direction this PR is taking the module is there anything you can recommend for me? :bowing_man:

veehaitch committed:
nixos/github-runners: set `DynamicUser=false` if `user` not `null`

veehaitch committed:
nixos/github-runners: add a `group` option to set the executing group

Similar to the `user` option, the added `group` option sets the group of
the executing process. If not `null`, it also sets `DynamicUser=false`.
In case `user` is set to `null` (the default), systemd would run the
service as root implicitly. As this is dangerous and most certainly not
what users want, we force them to set `user = "root"` explicitly if
that's really their intention. That's achieved through an assertion.

veehaitch commented:
> unfortunate for me... but this PR breaks my workflow 😞
> 
> the `systemd` hardening in this module just doesn't work for my usage and there isn't an easy way to undo it with this PR currently [this](https://github.com/aanderse/nix/blob/main/nixos/modules/github-runner/default.nix) is what i'm doing... though what i would really love is a way to simply strip out the `systemd` hardening in this module and then i wouldn't need to resort to the hacks i use in the link provided
> 
> @veehaitch given the direction this PR is taking the module is there anything you can recommend for me? 🙇‍♂️

With the latest version of this PR, you should be able to use `services.github-runners` directly without any hacks; `serviceOverrides` should do the job.

aanderse commented:
> `serviceOverrides` should do the job.

with `serviceOverrides` i need to figure out what all the default values are as well as keep a close eye on every PR to this module so i can defensively block new options as they are added

but i guess that is what i will need to do - ok, thanks

veehaitch commented:
I see. I'm not against having a new option which controls the default hardening options as long as those are enabled by default. That's probably another PR though 🙂

aanderse commented:
> I see. I'm not against having a new option which controls the default hardening options as long as those are enabled by default. That's probably another PR though 🙂

it's really more of a nixos in general thing, not specific to this module. the changes to this module just make my workaround ineffective... but that's not your problem.

thanks for taking such great care of this package and module - it is very helpful to me 👍
